### ISSUE TITLE
[help] hibernation support

### ISSUE THREAD

**Issue by julie-de-ville** [desultory/ugrd#82]
Is hibernation resume support included automatically? I don't see a module for it, and my system supports hibernation, so I am checking to see if this is the cause.

**Comment by desultory** [comment]
It does not currently have hibernation support, but this shouldn't be very hard to add.

Have you tried using the "resume=" kernel command line arg? I have not tested this but assumed it could be used to resume off of unencrypted swap.

**Comment by julie-de-ville** [comment]
That would be great, I found it much better than dracut. I have the resume parameter in grub.

**Comment by desultory** [comment]
Do you have encrypted swap? I think it shouldn't be too hard to add support for simple resuming, I'm just not sure why the builtin kernel parameter doesn't work alone. Maybe it ignores that option if an initrd is used?

**Comment by desultory** [comment]
https://wiki.gentoo.org/wiki/Custom_Initramfs/Hibernation

I've hesitated to add support in the initrd as there are many things which can go wrong. I think passing the supplied info to /sys/power/resume could be enough.

**Comment by desultory** [comment]
https://github.com/desultory/ugrd/commit/0175133404709c34686c811012d900aca4107077
I'm not sure about this, but I think it may be a reasonable start? resume= expects a device path, I'm not sure if it makes sense to resolve a UUID.

**Comment by desultory** [comment]
Right now, it enters a fail state if a resume partition is passed and it fails to resume, this means it won't normally boot. I'm not sure how much to consider those warnings about data loss. If it hibernated, and you reboot without considering the saved state, there could potentially be serious data loss, similar to if you did a hard shutdown. I think most systems just attempt to resume from swap if possible, but continue if not. I'm going to check/test a bit more

**Comment by desultory** [comment]
> That would be great, I found it much better than dracut. I have the resume parameter in grub.

did you manually add the parameter? I think for the sake of safety, I will be forcing resume attempts if resume= is set. It's potentially very dangerous to start a system fresh if it expects to return from a hibernation state.

**Comment by julie-de-ville** [comment]
Yes, I manually added it to grub. I booted twice in that manner, but I haven't attempted resuming again since I have narrowed it down to initramfs. That is a good idea.

**Comment by desultory** [comment]
> Yes, I manually added it to grub. I booted twice in that manner, but I haven't attempted resuming again since I have narrowed it down to initramfs. That is a good idea.

yeah it's probably not safe to resume right now, you could have a bit of data loss each time as it expects to later resume from the current ram state.

As far as I know, there is no way to know if a system _should_ resume at boot time, other than the passed kernel cmdline parameters. It's safest to prevent booting if that was passed but cannot be performed.

**Comment by julie-de-ville** [comment]
As for path/uuid, I have my resume parameter set to the mapped, decrypted luks partition, at /dev/mapper/gentoo-root. Btw, I am using btrfs on an encrypted luks partition, and for S5 I suspend to a swapfile on the root subvolume, if that helps.

**Comment by desultory** [comment]
> As for path/uuid, I have my resume parameter set to the mapped, decrypted luks partition, at /dev/mapper/gentoo-root. Btw, I am using btrfs on an encrypted luks partition, and for S5 I suspend to a swapfile on the root subvolume, if that helps. Also, I had to set the resume_offset as a parameter as well, since I am using a swapfile.

resume should be set to your swap partition, the support I just added only supports plain swap, I may add support for encrypted swap too.

As it is, it will boot normally if it cant mount using the provided resume path (using a partuuid is best), if you set `safe_resume = true` in the config, it will force the resume to be used or fail to boot.

**Comment by julie-de-ville** [comment]
Oh I see, my swap is encrypted so I don't think I will be able to test it safely.

**Comment by desultory** [comment]
this isn't really the best UX, but maybe you could take advantage of the fact that it fails, and then choose whether you want to manually run "crypt_init" which will run the cryptsetup unlock procedure or tell it to ignore resuming. Then you can exit the recovery shell, and on the second pass it will see the resume source and use that.

**Comment by julie-de-ville** [comment]
Thank you, how would I go about doing that? If you could point me in the right direction, that's be great. The resume offset is correct; if the offset is not calculated correctly, it will not hibernate at all, and now it hibernates but doesn't resume.

**Comment by desultory** [comment]
> Thank you, how would I go about doing that? If you could point me in the right direction, that's be great. The resume offset is correct; if the offset is not calculated correctly, it will not hibernate at all, and now it hibernates but doesn't resume.

If you add the `recovery` kernel cmdline arg, when it fails it should open a bash shell, from there you can try to do things manually and see what works. `crypt_init` is a function in ugrd which should run the luks procedures for your device. if you can do that, it may work. `handle_resume` should do the whole procedure to resume based on your kernel command line, but you could try to manually echo something to `/sys/power/resume`

**Comment by julie-de-ville** [comment]
Hi, I didn't see your last reply. Would it be possible to get an initramfs to resume from encrypted hibernation using those cmdline args, or would I have to do it manually each time?

**Comment by desultory** [comment]
> Hi, I didn't see your last reply. Would it be possible to get an initramfs to resume from encrypted hibernation using those cmdline args, or would I have to do it manually each time?

A user has made a patch which seems to make it work, I'm considering adding it but still unsure about safety, see: https://www.kernel.org/doc/html/latest/power/swsusp.html

I don't think it should be an issue since the mount is made after the resume stage, but I'm not sure how this will work with other storage which is mounted, such as a device for keyfiles or headers.

```
diff --git a/src/ugrd/fs/resume.py b/src/ugrd/fs/resume.py
index 38f5ddd..9255d9c 100644
--- a/src/ugrd/fs/resume.py
+++ b/src/ugrd/fs/resume.py
@@ -1,7 +1,8 @@
 __version__ = "0.4.0"
 
+from zenlib.util import contains, unset
 
-def handle_resume(self) -> None:
+def _resume(self) -> None:
     """Returns a bash script handling resume from hibernation.
     Checks that /sys/power/resume is writable, resume= is set, and noresume is not set, if so,
     checks if PARTUUID= is in the resume var, and tries to use blkid to find the resume device.
@@ -17,7 +18,7 @@ def handle_resume(self) -> None:
     return [
         "resumeval=$(readvar resume)",  # read the cmdline resume var
         'if ! check_var noresume && [ -n "$resumeval" ] && [ -w /sys/power/resume ]; then',
-        '    if echo "$resumeval" | grep -q "PARTUUID="; then',  # resolve partuuid to device
+        '    if echo "$resumeval" | grep -q -E "^PARTUUID=|^UUID="; then',  # resolve partuuid or uuid to device
         '        resume=$(blkid -t "$resumeval" -o device)',
         "    else",
         "        resume=$resumeval",
@@ -35,3 +36,11 @@ def handle_resume(self) -> None:
         "    fi",
         "fi",
     ]
+
+@unset('late_resume')
+def handle_resume(self) -> None:
+    return _resume(self)
+
+@contains('late_resume', "Using late resume.")
+def late_resume(self) -> None:
+    return _resume(self)
\ No newline at end of file
diff --git a/src/ugrd/fs/resume.toml b/src/ugrd/fs/resume.toml
index c4732dd..fa74744 100644
--- a/src/ugrd/fs/resume.toml
+++ b/src/ugrd/fs/resume.toml
@@ -2,3 +2,9 @@ cmdline_strings = [ "resume" ]
 
 [imports.init_early] 
 "ugrd.fs.resume" = [ "handle_resume" ]
+
+[imports.init_premount]
+"ugrd.fs.resume" = [ "late_resume" ]
+
+[custom_parameters]
+late_resume = "bool"
\ No newline at end of file
```

**Comment by linuxandos2** [comment]
> > Oh I see, my swap is encrypted so I don't think I will be able to test it safely, though I would really like to get it working, so if there is anything I can do to help lmk. I tried to get it working with dracut by adding the crypt and resume modules, and including /etc/crypttab, but it hung on a black screen with a spinning wheel.
> 
> Using encrypted swap is somewhat complex. I'd have to add a new method specifically for opening that, which can run first. The real tricky part is that would likely need to attempt to run on every boot. When you're booting fresh, that will just be a waste of time because it will not be able to resume.
> 
> you mentioned `/dev/mapper/gentoo-root` as being your resume device, is that your root partition? Do you have a separate partition that is luks encrypted, or are you using lvm? resuming from swap files is especially difficult because the file offset on the disk must be set.
> 
> > As for path/uuid, I have my resume parameter set to the mapped, decrypted luks partition, at /dev/mapper/gentoo-root. Btw, I am using btrfs on an encrypted luks partition, and for S5 I suspend to a swapfile on the root subvolume, if that helps. Also, I had to set the resume_offset as a parameter as well, since I am using a swapfile.
> 
> Are you sure the resume_offset you found is correct? That is my only assumption why dracut may fail, unless it doesn't properly support hibernation from luks devices. Did it ever ask for the key for your root device? I really hesitate to even open luks devices because I'm not sure if opening them has any chance at writing anything. If you touch storage devices at all between hibernation and resuming, that be may harmful.

Changing the order for handle_resume cfr crypt_init so that handle_resume commes after crypt_init is enough to make it work.
like 
```
[import_order.after]
handle_resume = "crypt_init"
```
in resume.toml

Tested with swap on encrypted partition.
Resume to ram and hibernate-resume works good.

**Comment by akira7095** [comment]
When the hibernation feature is enabled, the following error appears during a normal startup (when not resuming from hibernation).
`*ERROR: Failed to resume from: /dev/mapper/cryptroot`
The startup process runs normally, but an error message appears.
Is there a way to fix this error message?

<details>
<summary>/etc/ugrd/config.toml</summary>
<pre>
<code>
# This config will decrypt the LUKS volume with uuid "fdf442da-0574-4531-98c7-55227a041f1d", mapping it to "/dev/mapper/root"
# Note: In most cases, the uuid of the LUKS volume will be detected from the root mount. It's set as an example here.
# It will attempt to mount the btrfs volume with label "rootfs" to /target_rootfs
# It will pull all current kernel modules from lspci -k results
# It will try to process the cmdline and mount the rootfs based on the root= parameter

late_resume = true

modules = [
  "ugrd.crypto.cryptsetup", # This is included by the gpg module
#  "ugrd.crypto.gpg", # This is included by the smartcard module
#  "ugrd.crypto.smartcard",
  "ugrd.fs.resume",
]

# Mask the parse_cmdline function to prevent it from running
# Can be applied to any function
#mask.init_pre = 'parse_cmdline'

# By default, all kmod categories are ignored. This can be disabled by category
# Modules can define these options if needed, and users should not need to set them
# In most cases, kmods in these categories are not needed to boot the system
#kmod_ignore_video = false
#kmod_ignore_sound = false
#kmod_ignore_network = false

# The initramfs will be built in /tmp/initramfs if "build_dir" is not specified
#out_dir = "/usr/src/initramfs"
#out_dir = "/boot"

# create device nodes in build dir, requires root
#make_nodes = true

# By default, xz compression is used. This can be disabled or set to zstd if the zstandard python module is installed
#cpio_compression = "zstd"  # Use zstd compression
#cpio_compression = false  # No compression

# Optionally supply a kernel version, uses the current kernel version if not specified
#kernel_version = "6.1.53-gentoo-dist"
# lspci based detection is usually helpful
kmod_autodetect_lspci = true
# lsmod detection can be helpful, but can bring in unnecessary modules
#kmod_autodetect_lsmod = true
# keyboard/input device detection is enabled by default, but can be disabled or adjusted
#kmod_autodetect_input = false
# If input detection is finding non-keyboard devices, the key count threshold can be adjusted
#keyboard_key_threshold = 50
# If no kernel modules are needed, or installed, no_kmod should be set
#no_kmod = true

# ugrd is designed for merged-usr systems. If using a split-usr system, automatic symlink creation can be disabled:
#merge_usr = false

# on systems using glibc in the userspace, libgcc is often required. If it's not necessary, pulling it can be diabled with:
#find_libgcc = false

# By default, ugrd will keep retrying to mount base mounts and the rootfs, unless space is pressed
# The default timeout between attempts is 1s, and the default number of retries is infinite
#mount_timeout = 0.25  # Wait 0.25 seconds between mount attempts, instead of 1s.
#mount_retries = 5  # Retry mounting the rootfs 5 times, if unset, it will retry indefinitely

# The location of the gpg public key, if using a smartcard and gpg encrypted keyfile
#sc_public_key = "/etc/ugrd/pubkey.gpg"

#auto_mounts = ['/boot']  # Automatically create mount config for /boot on the host

# Optionally specify the root mountpoint information manually, it should be detected automatically
# Whatever options are set manually will be validated against the current root mount
#[mounts.root]
# The label of the root filesystem, a uuid or partuuid could be used instead
#label = "rootfs"

# This is necessary if the keyfile is on another device.
# The keyfile could be added as a dependency. if it's on the same device.
#[mounts.boot]
#type = "vfat"
#uuid = "4886-A631"

# If the root is a LUKS volume, specify the LUKS volume information
# The name of the LUKS volume should match the name of the device in /dev/mapper
#[cryptsetup.root]
# The UUID of the encrypted volume, mounted at /dev/mapper/root, should be detected automatically
#uuid = "fdf442da-0574-4531-98c7-55227a041f1d"

# If a GPG encrypted key file is used, the key_type must be set to gpg.
# If using a plain keyfile, this can be left unset.
#key_type = "gpg"
# The key file must either be included as a dependency, or it must be under an external mountpoint.
# Key file under /boot, which is configured as an auto_mount
#key_file = "/boot/rootfs.luks.gpg"
# Key file included as a dependency
#key_file = "/etc/ugrd/rootfs.luks.gpg"
#include_key = true
# If the key file is on storage not accessible to the build system, validation can be disabled:
#validate_key = false
# If key unlocking fails, attempt a password unlock:
#try_nokey = true

# If using detached headers, the header file can be treated similarly to the key file
#header_file = "/boot/rootfs.luks.header"
#header_file = "/etc/ugrd/rootfs.luks.header"
#include_header = true
#validate_header = false

no_fsck = true
</code>
</pre>
</details>

**Comment by desultory** [comment]
> When the hibernation feature is enabled, the following error appears during a normal startup (when not resuming from hibernation). `*ERROR: Failed to resume from: /dev/mapper/cryptroot` The startup process runs normally, but an error message appears. Is there a way to fix this error message?
> /etc/ugrd/config.toml

im not sure, the concern is that the script (i think) should always attempt to consider resume= args if passed, so if your bootloader is passing them and there is nothing to resume from, it can only know that once it tries.
